text/0000-range_redesign.md

```diff
@@ -0,0 +1,67 @@
+- Feature Name: range_redesign
+- Start Date: 31/03/2019
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+A Range reimplementation that includes the features from collections.Reverse together with a more comprehensive and coherent API. A reference implementation can be found here: https://github.com/chobeat/pony-range-proposal
+
+# Motivation
+
+The existence of Range and Reverse as separate classes, each one with their own limitations, to cover a use case that in most languages is addressed by a single API highlights the possibility for an improvement. Having a single, well-tested implementation to generate numeric ranges will improve the quality of life of new developers and possibly reduce unexpected behaviors for a core piece of the stdlib.
+For example in the existing implementation it is possible to generate infinite ranges without any explicit control. 
+
+In addition to that, the current implementation doesn't allow to express a range spanning over a whole data type since it makes assumptions on the exclusiveness of the right bound.
+
+# Detailed design
+
+The new implementation should fullfil the following design principles: 
+
+* be able to generate all the possible ranges, increasing and decreasing, for all the data types, without incurring in overflow errors.
```

Comment:
how does the reference implementation achieve this goal?

Reply:
First of all there are explicit tests to verify this. I would say that the main problem before directly addressing this goal was that some parts of the implementation (very similar to the existing one), when the bounds was the max_value or min_value of some data type, were overflowing when checking if the bound was reached, basically creating an infinite range. Now this is handled explicitely: when the next value would reach the bound and if that bound would have overflow problems, then we just end the range there.